### ISSUE TITLE
Text from old xopp file does not show up in new XOPP v1.1.0

### ISSUE THREAD

**Issue by axsvl77** [xournalpp/xournalpp#3371]
**Affects versions :**
 - OS: Windows 10
 - libgtk 3.24.29
 - Version of Xournal++: 1.10 9d1277de
 - Official release

**Describe the bug**
I save a file with Xounalpp in Jan 2021. I installed the latest version of xopp. Now, the text looks like little boxes.

**To Reproduce**
1. Save with old version of Xounalpp
2. Install new version of Xournalpp
3. Open the old file

**Expected behavior**
There text should render properly, and be editable.

**Screenshots of Problem**
Side by side, an old pdf export, and the current file

**Additional context**
Add any other information or context about the problem here.

**Comment by LittleHuba** [comment]
Could you check which font you used in the document? Either in the PDF or by decompressing the xopp file? Such boxes are normally used by the renderer if the used font does not exist on the system.

**Comment by axsvl77** [comment]
It says "Times New Roman Normal". I found this in the older version of XPP (1.0.18)

**Comment by LittleHuba** [comment]
Can you confirm that in the newer version this font is still available?

**Comment by axsvl77** [comment]
@LittleHuba 

It says the font "Times New Roman Normal" is on the list but doesn't seem to work.
See images:

![xournalpp_uKWK0dYjq8](https://user-images.githubusercontent.com/11460222/132440458-91fb811f-d855-4da5-91a4-9a6f6e0e29ee.png)

When I switch to Times New Roman Bold Italic, is works again:

![2021_09_07_23_10_54](https://user-images.githubusercontent.com/11460222/132440466-570612fe-e2d0-471f-b4cf-b619bad08e49.png)

**Comment by LittleHuba** [comment]
Okay. Then something with this specific font is problematic.

You can either send me the file via Gutter and I'll fix it for you or you can do so manually by changing the font in all the text boxes.

Could you run Xournal++ from the console and check if there are any log messages concerning fonts?

**Comment by LittleHuba** [comment]
If you send me a single file where you modify one textbook to a different font that works for you, I'll write you a small script that modifies all the text boxes in all the files.

**Comment by LittleHuba** [comment]
You can install either Git for Windows (https://gitforwindows.org) or MSYS2 (https://www.msys2.org) and use the git bash or mingw64 bash to start Xournal++. There you get the debug output from Xournal++.

**Comment by personalizedrefrigerator** [comment]
If you have some python experience and want to write a script that changes the font yourself:
<details>

`.xopp` files are [GZipped](https://en.wikipedia.org/wiki/Gzip) XML files. Python has a built-in [library for reading and writing `gzip`ped files.](https://docs.python.org/3/library/gzip.html#module-gzip).

<details><summary>For example, the following is a very simple `.xopp` file:</summary>

![XOPP file with a single line of text](https://user-images.githubusercontent.com/46334387/132643727-cbeb47fd-0cf4-40e9-8d5e-70d93fcfd2a0.png)
![](https://user-images.githubusercontent.com/46334387/132644444-d148b547-4978-4295-a3a3-72b4fc5db6dc.png)

</details>

In the example, notice that the text was labeled: `font="Serif"`, and not `font="Serif Regular"`. Make sure you know what this looks like in your `.xopp` files! A `python` script to do this might look like this:
```python
import gzip

PATH_TO_EXAMPLE_XOPP = "C:\\\\Users\\You\\Absolute\\Path\\To\\An\\XOPP\\File"

# 'rt' stands for 'readonly textmode'.
with gzip.open(PATH_TO_EXAMPLE_XOPP, 'rt') as file:
    for line in file.readlines():
      if "font" in line:
        # Print out every line that contains 'font'.
        print(line)
```

As such, one might write a script that looks similar to this (**which I have not tested on Windows**):
```python
import gzip
from pathlib import Path

# We'll use shutil for file-copying. See https://docs.python.org/3/library/shutil.html
from shutil import copyfile

# 'rt' stands for 'read textmode'. We'll give this to `gzip.open`.
OPEN_IN_READONLY_TEXTMODE = "rt"
OPEN_IN_WRITEONLY_TEXTMODE = "wt"
OLD_FONT_NAME = "Times New Roman Normal"
NEW_FONT_NAME = "Times New Roman Bold"
TARGET_DIRECTORY = Path("C:\\\\Users\\Path\\To\\Your\\XOPP\\Files")

def fix_xopp_file(path):
  # First, make a backup!
  # We'll read from the backup, but won't delete it in case something goes wrong!
  backup_filepath = path + ".backup"
  copyfile(path, backup_filepath)

  # While we could set output_filepath to path, if something else goes wrong,
  # we'll still have the original '.xopp' files.
  # If you've tested this script on some example files, feel free to replace this with
  # output_filepath = path.
  output_filepath = path + ".fixedup.xopp"
  
  # See https://stackoverflow.com/questions/31028815/how-to-unzip-gz-file-using-python
  # and https://docs.python.org/3/library/gzip.html#module-gzip
  with gzip.open(backup_filepath, OPEN_IN_READONLY_TEXTMODE) as input_file:
    with gzip.open(output_filepath, OPEN_IN_WRITEONLY_TEXTMODE) as output_file:
      for line in input_file.readlines():
        updated_line = line.replace(OLD_FONT_NAME, NEW_FONT_NAME)
        output_file.write(updated_line + "\n")

# The `list` here is probably important: We'll be creating new .xopp files and we don't want
# these new files to be included in the iteration.
xopp_files = list(TARGET_DIRECTORY.glob("*.xopp"))

for path in xopp_files:
  # Convert from a Path object to a string
  # Note that 'path.resolve' gives us an absolute path.
  pathname = str(path.resolve())

  if not pathname.endswith('.fixedup.xopp'):
    print("Fixing %s..." % path)
    fix_xopp_file(pathname)
    print("  Done!")


```

Note that one issue with this method is that any other references to the font get replaced, as well. The example from before became:
![](https://user-images.githubusercontent.com/46334387/132646130-436e4950-dd08-4086-a498-53fdb8a2b050.png)


</details>

**Comment by LittleHuba** [comment]
Nice script but please do not only search for the font itself but use the string`font="FONT"`. Otherwise, you might edit more than you want to.

**Comment by axsvl77** [comment]
@personalizedrefrigerator Thanks, I will look into that.

**Comment by axsvl77** [comment]
@LittleHuba I am enjoying learning about this. So I have 
(1) installed Git for windows, 
(2) started Git Bash /  mingw64  
(3) figured out how to start xournalpp.exe from mignw64

What do I do next to get the " get the debug output from Xournal++"?

**Comment by axsvl77** [comment]
FYI, I installed it on another system and the issue is bigger here:
![xournalpp_olrEJIhoAC](https://user-images.githubusercontent.com/11460222/132795625-f87344ab-0656-438f-8c22-88c403c8deb4.png)

What's gutter?

**Comment by LittleHuba** [comment]
> @LittleHuba I am enjoying learning about this. So I have 
> (1) installed Git for windows, 
> (2) started Git Bash /  mingw64  
> (3) figured out how to start xournalpp.exe from mignw64
> 
> What do I do next to get the " get the debug output from Xournal++"?

When you start Xournal++ from a git bash console you should see some lines printed in this console. Those lines are the ones I'm interested in.

**Comment by LittleHuba** [comment]
> FYI, I installed it on another system and the issue is bigger here:
> ![xournalpp_olrEJIhoAC](https://user-images.githubusercontent.com/11460222/132795625-f87344ab-0656-438f-8c22-88c403c8deb4.png)
> 
> What's gutter?
> 

Wow that is around 100% worse... Can you also provide the same console output from this system? What font are you using there as UI font in Windows? And what version of Xournal++ did you use?

**Comment by axsvl77** [comment]
@LittleHuba 

(1) This is the newest version, 1.1.0. I downloaded it yesterday. Note that 1.0.18 didn't have this problem on this system. 

(2) The menu and prompt font error only happens when I open the file by double clicking on the file in the folder. When I open Xournal++ from the start menu then open the same file via the Xournal++ open dialog, only the file fonts are broken. Same behavior on both systems.

Sometimes it works just fine; it isn't at all consistent. As I am using really like Xournal++ during live lectures, this is fairly irritating.

(3) I am using the default font, as it turns out that changing the system font in Windows [is fairly involved](https://www.howtogeek.com/716407/how-to-change-the-default-system-font-on-windows-10/). It might be Segoe UI.

(4) I am happy to provide a console output, but it doesn't have much information. Should I use a different command to get error logs?

![mintty_qWfB1SQJbz](https://user-images.githubusercontent.com/11460222/132843295-f1cae447-7686-454f-9b39-f544bfd168de.png)

**Comment by LittleHuba** [comment]
You must start Xournal++ with `xournalpp` instead of `start xournalpp`.

**Comment by axsvl77** [comment]
It doesn't work
![mintty_xQdZGr9wu3](https://user-images.githubusercontent.com/11460222/132849757-7de6103f-acaa-4f55-aca6-19c7058daeda.png)

**Comment by personalizedrefrigerator** [comment]
> It doesn't work
> ![mintty_xQdZGr9wu3](https://user-images.githubusercontent.com/11460222/132849757-7de6103f-acaa-4f55-aca6-19c7058daeda.png)

Try ./xournalpp and ./xournalpp.exe (for why, see https://stackoverflow.com/questions/6331075/why-do-you-need-dot-slash-before-executable-or-script-name-to-run-it-in-bas)
First, though, make sure you're in the same directory as xournalpp!

**Comment by axsvl77** [comment]
@LittleHuba @personalizedrefrigerator 

Here is the text from migw64:
(xournalpp.exe:19264): GLib-GIO-WARNING **: 10:28:43.940: win32 session dbus binary not found: C:\Program Files\Xournal++\bin\\gdbus.exe
** Message: 10:28:44.112: TEXTDOMAINDIR = (NULL), PACKAGE_LOCALE_DIR = ../share/locale/, chosen directory = ../share/locale/

** (xournalpp.exe:19264): WARNING **: 10:28:44.120: XournalMain: System default locale could not be set.
 - Caused by: locale::facet::_S_create_c_locale name not valid
 - Note that it is not supported to set the locale using mingw-w64 on windows.
 - This could be solved by compiling xournalpp with msvc
** Message: 10:28:44.249: Plugin "MigrateFontSizes" UI initialized
** Message: 10:29:26.272: lastOpenPath set
** Message: 10:29:26.281: file: C:\Users\slouis\Dropbox\OtherDropbox\Faculty\ECE2005\Lectures\troubleshooting\Topic2 Examples Solved.xopp

![signal-2021-09-10-105338_001](https://user-images.githubusercontent.com/11460222/132873761-a6f932b4-6e95-435d-8916-d5ac10bdf093.png)


Did this provide anything useful?

**Comment by LittleHuba** [comment]
This one seems to be problematic...

> GLib-GIO-WARNING **: 10:28:43.940: win32 session dbus binary not found: C:\Program Files\Xournal++\bin\gdbus.exe

Could you check if this file does really not exist?

**Comment by axsvl77** [comment]
@LittleHuba I looked on both computers, and it is not there. I did a search also through the file system; not anywhere on either machine.

**Comment by LittleHuba** [comment]
Okay thanks!
I'll get back to this ticket on Sunday or Monday. Hopefully we only have to add this executable to the packaging script in the pipeline.

If that doesn't fix it, we'll have to make a git bisect to find the change that broke this for you.

**Comment by axsvl77** [comment]
@LittleHuba Hi! Don't want you to forget this; curious to see how the process will work of getting it fixed.

**Comment by LittleHuba** [comment]
Our Windows Pipeline is currently broken. Once we have it running again, I'll add the executable and provide you with a link to the installer.

**Comment by LittleHuba** [comment]
Sorry for the long wait. Now that our pipelines are working again, I got around to trying a patch for this issue. I can unfortunately not reproduce your bug, so You'll have to test the fix for me.

The pipelines are just now building a patched installer. The links will appear in about 15 minutes via a comment in #3426. Could you please give it a try once it is built?

**Comment by axsvl77** [comment]
@LittleHuba Thanks for doing this patch. I did the following:
1. Confirmed I still had this problem on the stable version of Xournal++ 
2. Uninstalled the stable version of Xournal++
3. Followed the link you have above, and then downloaded the "windows.zip" patch file.
4. Unzipped the file with winrar, then installed the patch on my computer

It is not 100% fixed, but it is better than before.

Case A:
![xournalpp_NLSTCEaF3I](https://user-images.githubusercontent.com/11460222/135945588-1b34b790-7c55-4ccf-92c9-2b3b49f3a24c.png)

With this, I opened the old file by right clicking on it, they choosing Xournal++. Good news is that the menus now have text instead of little boxes. Bad news is that the text in my document retains the little boxes. I can, however, go into the document and change the font 1 by 1 so that it is visible.

Case B:
When I open Xournal++ first, then open the same file, everything works perfectly.

The version information looks like this:

![firefox_Vfrn5XPfXU](https://user-images.githubusercontent.com/11460222/135945829-50d5c804-b892-467e-999a-ee1bf66e3d36.png)

So, as discussed in "Case A" a small bug remains. I am interested in helping to solve this problem, when you have the time.

**Comment by LittleHuba** [comment]
Thanks for the report. Could you provide the console output once more? There may be something else that is reported now.

**Comment by ikanakova** [comment]
Followed the link you have above, downloaded the "windows.zip" patch file and reinstalled Xournal++, I still have a problem #3433.
Uninstalled Xournal++, install patch file and the error persists.
I think #3433 is not related to this issue because there is a problem with the GUI (the appearance of the application).

**Comment by LittleHuba** [comment]
@axsvl77 should #3426 not fix this issue, please also try the installer from #3477.

**Comment by axsvl77** [comment]
It says file type not supported

**Comment by LittleHuba** [comment]
> It still has the problem. Trying the next one.
> This has the same problem.

Hmm was worth a shot... I'll have to dig deeper. Most likely I'll give you another installer that dumps a ton of debug messages. Will take a little bit to create such a version.

**Comment by LittleHuba** [comment]
Bad news... I can still not replicate your issue, even with this test file. So it must be something specific to your setup. Please remove the settings file of Xournal++, so we can make sure that there is no misconfigruation. (See this page for where to find this file https://xournalpp.github.io/guide/file-locations/). Xournal++ may not be open while you remove the file. Once you removed it, Xournal++ will generate a default settings file on the next start. If that doesn't help you, I'll start with the debug enabled custom installer.

**Comment by cold-bb** [comment]
Any progress? I have the same problem when I open the xopp file from anywhere that is NOT the Xournal++ open file dialogue. e.g. when I open the file directly from desktop. Windows 10 Pro (64 bit) (version: 21H1)

**Comment by LittleHuba** [comment]
My free time is quite limited currently. I haven't forgotten about this though and will get back to it once I find some free time. Anybody with development experience is of course free to take over. After all the faster this is fixed, the better.

**Comment by LittleHuba** [comment]
I just had a little bit of time and could dig some more. This seems to be an issue within Pango. From the general behavior of the bug (e.g. intermittency), I assume it is some kind of racing condition. Pango seems to have some problems with concurrent accesses which is bad news for us since our rendering engine heavily relies on concurrency. 

That is unfortunately as far as I got. Pango is not very cooperative with debug messages that would tell you what went wrong.

This bug should be elevated upstream to the developers of Pango.

What we can try until they have a fix available:
1. Downgrade Pango on CI and hope that the previous release does not have this bug (most likely requires a manual build of Pango)
2. Sequentialize the rendering of text views temporarily (potentially huge impact on performance and doesn't help with GUI elements)

**Comment by LittleHuba** [comment]
A possible fix seems to be setting the environment variable `PANGOCAIRO_BACKEND=fc` which seems to use a different backend that is thread-safe. I made a quick attempt at a fix in #3534. Installers are built while I'm writing this. Please try them once they are available (~15 min from now).

**Comment by LittleHuba** [comment]
Build succeeded this time and I can't reproduce the issue any longer. Please try it out as well.

**Comment by cold-bb** [comment]
Where does one find those installers?

**Comment by cold-bb** [comment]
Thanks. I have installed it and the issue seems resolved

**Comment by bhennion** [comment]
Fixed via #3426 via #3599
